profiler/profile_test.go

```diff
@@ -333,6 +348,277 @@ main.main()
+		// not trying assert the quality of the profiler beyond the presence of the
+		// right labels.
+		minCPUDuration = cpuDuration / 10
+	)
+
+	assertCommon := func(t *testing.T, app *apmTestApp, res apmTestResponse) {
+		require.True(t, validSpanID(res.SpandID))
+		require.True(t, validSpanID(res.LocalRootSpanID))
+		require.Greater(t, app.CPUTime(t), minCPUDuration)
+		require.Greater(t, app.LabelsCPUTime(t, apmTestCustomLabels), minCPUDuration)
+	}
+
+	t.Run("none", func(t *testing.T) {
+		app := startAPMTestApp(t)
+		defer app.Stop(t)
+
+		res := app.Request(t, cpuDuration)
+		assertCommon(t, app, res)
+		require.Zero(t, app.LabelCPUTime(t, intprof.SpanID, res.SpandID))
+		require.Zero(t, app.LabelCPUTime(t, intprof.LocalRootSpanID, res.LocalRootSpanID))
+		require.Zero(t, app.LabelCPUTime(t, intprof.TraceEndpoint, endpoint))
+	})
+
+	t.Run("endpoint-filtering", func(t *testing.T) {
+		app := startAPMTestApp(t, tracer.WithProfilerEndpoints(true))
+		defer app.Stop(t)
+
+		res := app.Request(t, cpuDuration)
+		assertCommon(t, app, res)
+		require.Zero(t, app.LabelCPUTime(t, intprof.SpanID, res.SpandID))
+		require.Zero(t, app.LabelCPUTime(t, intprof.LocalRootSpanID, res.LocalRootSpanID))
+		require.Greater(t, app.LabelCPUTime(t, intprof.TraceEndpoint, endpoint), minCPUDuration)
+	})
+
+	t.Run("code-hotspots", func(t *testing.T) {
+		app := startAPMTestApp(t, tracer.WithProfilerCodeHotspots(true))
+		defer app.Stop(t)
+
+		res := app.Request(t, cpuDuration)
+		assertCommon(t, app, res)
+		require.Greater(t, app.LabelsCPUTime(t, map[string]string{
+			intprof.SpanID:          res.SpandID,
+			intprof.LocalRootSpanID: res.LocalRootSpanID,
+		}), minCPUDuration)
+		require.Zero(t, app.LabelCPUTime(t, intprof.TraceEndpoint, endpoint))
+	})
+
+	t.Run("all", func(t *testing.T) {
+		app := startAPMTestApp(t, tracer.WithProfilerEndpoints(true), tracer.WithProfilerCodeHotspots(true))
+		defer app.Stop(t)
+
+		res := app.Request(t, cpuDuration)
+		assertCommon(t, app, res)
+		require.Greater(t, app.LabelsCPUTime(t, map[string]string{
+			intprof.SpanID:          res.SpandID,
+			intprof.LocalRootSpanID: res.LocalRootSpanID,
+			intprof.TraceEndpoint:   endpoint,
+		}), minCPUDuration)
+	})
+
+	t.Run("none-child-of", func(t *testing.T) {
+		app := startAPMTestApp(t)
+		defer app.Stop(t)
+		app.UseWithChild(true)
+
+		res := app.Request(t, cpuDuration)
+		assertCommon(t, app, res)
+		require.Zero(t, app.LabelCPUTime(t, intprof.SpanID, res.SpandID))
+		require.Zero(t, app.LabelCPUTime(t, intprof.LocalRootSpanID, res.LocalRootSpanID))
+		require.Zero(t, app.LabelCPUTime(t, intprof.TraceEndpoint, endpoint))
+	})
+
+	t.Run("all-child-of", func(t *testing.T) {
+		app := startAPMTestApp(t, tracer.WithProfilerEndpoints(true), tracer.WithProfilerCodeHotspots(true))
+		defer app.Stop(t)
+		app.UseWithChild(true)
+
+		res := app.Request(t, cpuDuration)
+		t.Skip("ChildOf does not propagate pprof ctx labels yet : /")
```

Comment:
@gbbr we might be able to fix this if we added a `Context() context.Context` to `ddtrace.SpanContext`, would you be opposed to this?

Reply:
I think that makes sense. I'm just concerned about a weird API call such as:
```go
span.Context().Context()
```
The first call returns a `SpanContext` and the next one returns a Go context. And, it would be the equivalent to `ContextWithSpan(context.Background(), span)`.

I guess what I'm realising by writing this is that it would not really have any use outside your internal use. But let's find a solution to this. What if we hide it? What if we add the `Context` method to the `spanContext`, which you can then check by doing something like:
```go
if v, ok := spanContext.(interface{ Context() context.Context }); ok {
    // use v.Context()
}
```
It's not great, but it fixes your problem. WDYT?

Reply:
Sounds reasonable! I'll try it out and push another commit if it works.